### ISSUE TITLE
convertSimpleArgument converts everything to string

### ISSUE THREAD

**Issue by br3w0r** [jackc/pgx#1551]
Could you please explain why `convertSimpleArgument` in `values.go` converts everything to `string`? If I understood correctly, it then breaks the result from `sanitize.SanitizeSQL` in `(*Conn).SanitizeForSimpleQuery`. I tried to make a simple test and the result is as expected – `int64` got converted to `string`
<img width="591" alt="image" src="https://user-images.githubusercontent.com/22412466/224779440-bd5c299e-85d3-49fb-869d-277232175137.png">
<img width="272" alt="image" src="https://user-images.githubusercontent.com/22412466/224779528-106b0558-a216-4165-a199-d17d24eee95b.png">

**Comment by jackc** [comment]
It's been a while so I don't remember the reason for sure. It could be a mistake. It also might be that it was not considered necessary since should always work thanks to PostgreSQL doing type conversion automatically. 

It's also simpler -- I took a quick stab at bypassing convertSimpleArgument for primitive types such as `int64` and `float64`. I quickly ran into some test failures because you can't rely on a float not needing quotes. `Infinity` and `NaN` complicate things.

I'm not against someone tightening this logic up if they want. But given that this path only occurs when parameters and results are in the text format I don't _think_ it makes any difference when executing queries.

**Comment by br3w0r** [comment]
Yes, it doesn't make any difference most of the time, but with `UPDATE ... FROM (VALUES (...))` statement it breaks the query as we don't know the concrete types in `VALUES`. Given an example for my test:

```go
func Test_test(t *testing.T) {
	conn, err := pgx.Connect(context.Background(), "user=postgres password=sample_pass host=127.0.0.1 port=57333 dbname=test-db sslmode=disable")
	assert.Nil(t, err)

	res, err := conn.SanitizeForSimpleQuery(
`
UPDATE
    test_table as t1
SET
    other_id = t2.other_id
FROM
    (
        VALUES
            (
                $1,
                $2
            )
    ) as t2(
        id,
        other_id
    )
WHERE
    t1.id = t2.id
`, int64(1), int64(123))

	fmt.Println(res)
}
```

With a table and data:
```sql
create table test_table(
	id int8 not null,
	other_id int8 not null
);

insert into test_table values (1, 111);
```

`SanitizeForSimpleQuery` produces a query:

```sql
UPDATE
    test_table as t1
SET
    other_id = t2.other_id
FROM
    (
        VALUES
            (
                '1',
                '123'
            )
    ) as t2(
        id,
        other_id
    )
WHERE
    t1.id = t2.id
 ```
 
 And this query passed to a DB gives an error:
 
 ```
SQL Error [42883]: ERROR: operator does not exist: bigint = text
Hint: No operator matches the given name and argument type(s). You might need to add explicit type casts.
 ```

**Comment by jackc** [comment]
> If you agree that it's a problem, I'd like to try fixing it this weekend

Please do!

**Comment by Se7ge** [comment]
@br3w0r Hi! Any progress with the issue?

**Comment by br3w0r** [comment]
@Se7ge the only fine solution I found is to revert back changes to this method and leave it the same as in v3: https://github.com/jackc/pgx/blob/a413de981978e4100d5736af7aaa2b392511612f/values.go#L17

**Comment by cruvie** [comment]
@jackc Hi, could you pls take some time have a look at https://github.com/go-gorm/gorm/issues/7584, things become worse with type alias with String() method

**Comment by cruvie** [comment]
my temp solution is, if this acceptable for pgx I will fire a pr
```go
func convertSimpleArgument(m *pgtype.Map, arg any) (any, error) {
	fieldValue := reflect.ValueOf(arg)
	switch fieldValue.Kind() {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		arg = fieldValue.Int()
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		arg = fieldValue.Uint()
	}
	buf, err := m.Encode(0, TextFormatCode, arg, []byte{})
	if err != nil {
		return nil, err
	}
	if buf == nil {
		return nil, nil
	}
	return string(buf), nil
}
```

**Comment by jackc** [comment]
@cruvie I don't think this approach can work. Some integer types have different encoding behavior that this would skip. See `TestIntervalCodec` for one example.

**Comment by cruvie** [comment]
@jackc I have fixed the tests, could you approval the workflow again?

**Comment by jackc** [comment]
@cruvie I approved the workflow.

But I took a quick glance at the code change and I don't think that it will work in all cases. Even if it does fix `TestIntervalCodec` it is entirely possible for a custom type with an underlying basic numeric type to define a pgx interface like `Int64Valuer`. I believe this new code would bypass the implemented pgx interface.

**Comment by cruvie** [comment]
@jackc thanks, I found another way to handle this issue, pls review the code, which could be clearer and more understandable

**Comment by jackc** [comment]
I don't think bypassing `Stringer` will work either. For a slightly contrived example, consider a PostgreSQL text column that is mapped to and from a Go `int` acting as an enum  in PostgreSQL.

For example:

| PostgreSQL | Go |
|:------------|----:|
| apple | 1 |
| orange | 2 |
| pear | 3 |

```go
type Fruit int

func (f Fruit) String() string {
...
}
```

I don't think that it is possible to use a short circuit based on underlying type without introducing the possibility of breaking behavior defined by a custom Codec or value type.

And even worse, after reviewing the existing pgx code and thinking some more I'm not even confident this issue *is* solvable. As mentioned in my original response over 2 years ago, even knowing the type is not sufficient to know whether the value must be quoted or not. The float8 `3.14` does not need to be quoted. But the float `'Infinity'` does.

Look at the example @br3w0r gave above with a `UPDATE ... FROM (VALUES (...))`. If that table needed a float, I don't think it would be possible to encode float `'Infinity'` without an explicit cast in the `VALUES` clause.

**Comment by F21** [comment]
I am also currently expriencing this problem where `uuid.UUID` from https://github.com/google/uuid is converted to `text`.

This is my table:
```sql
CREATE TABLE "test" (
    id UUID NOT NULL,
    tid UUID NOT NULL,
    rid TEXT NOT NULL,

    PRIMARY KEY (id, tid, rid)
)
```

This is my query: 
```sql
SELECT id, tid, rid FROM "test" WHERE id = $1 AND (tid, rid ) !=ALL(VALUES($2,$3),($4,$5))
```

The query fails for parameters `$2` and `$4`, even when the parameters are `uuid.UUID`: 
```go
rows, err := pool.Query(ctx, query, uuid.New(), uuid.New(), 'test1', uuid.New(), 'test2')
```
This is the error received: `ERROR: operator does not exist: uuid <> text (SQLSTATE 42883)`

The error does not happen if `$2` and `$4` are cast to `::uuid`. I think it would be much nicer DX if pgx is able to set the correct OIDs for those parameters. While it's possible to cast those parameters, it's a bit more troublesome and complicates things as I am generating the `($2,$3),($4,$5)` dynamically based on the number of inputs.
